### ISSUE TITLE
passTLSCert passes server instead of client certificate to the backend

### ISSUE THREAD

**Issue by steffkes** [containous/traefik#3052]
given that [the docs explain](https://docs.traefik.io/basics/) that

> You can also optionally enable `passTLSCert` to forward TLS Client certificates to the backend.

i'm trying to use traefik as ssl-proxy and a simple golang https server as dummy-endpoint to demonstrate the problem. as far as i'm aware there is no other tool that allows passing an entire client certificate forward .. i'm not even sure this could work as advertised?

i've created my own ca, a server and a client certificate signed from it (nothing special) - the whole thing is running in a docker compose environment:

```yaml 
version: '3.3'
services:

  proxy:
    image: traefik:1.5.4-alpine
    volumes:
      - .:/app
      - ./traefik.toml:/etc/traefik/traefik.toml:ro
    ports:
      - '127.0.0.1:5443:443'
      - '127.0.0.1:5088:8080'

  server:
    image: golang:1.10.0-alpine3.7
    command: sh -c 'go get -d -v ./... && go install -v ./... && app'
    working_dir: /go/src/app
    volumes:
      - .:/go/src/app
      - .:/app
    ports:
      - '127.0.0.1:5447:8080'
```

the traefik configuration in question:

```toml
logLevel = "DEBUG"
defaultEntryPoints = ["https"]
debug = true
RootCAs = ["/app/certs/ca/crt.pem"]

[api]
  [api.statistics]

[file]
  watch = true

[entryPoints]

  [entryPoints.https]
  address = ":443"

    [entryPoints.https.tls]
      
      [entryPoints.https.tls.ClientCA]
      files = ["/app/certs/ca/crt.pem"]
      optional = false

      [[entryPoints.https.tls.certificates]]
      certFile = "/app/certs/server/crt.pem"
      keyFile = "/app/certs/server/key.pem"

[frontends]

  [frontends.frontend-server]
    backend = "backend-server"
    passHostHeader = true
    passTLSCert = true

    [frontends.frontend-server.routes]
      [frontends.frontend-server.routes.route0]
        rule = "Host: server"

[backends]

  [backends.backend-server]
    [backends.backend-server.servers]
      [backends.backend-server.servers.server0]
        url = "https://server:8080"
``` 
the dummy endpoint's code:

```go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"io/ioutil"
	"net/http"
)

func main() {
	caCert, _ := ioutil.ReadFile("/app/certs/ca/crt.pem")
	caPool := x509.NewCertPool()
	caPool.AppendCertsFromPEM(caCert)

	tlsConfig := &tls.Config{
		ClientAuth: tls.RequireAndVerifyClientCert,
		ClientCAs:  caPool,
	}
	tlsConfig.BuildNameToCertificate()

	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
		fmt.Fprintf(
			w,
			"cert/subject: %v (%d)\n",
			req.TLS.PeerCertificates[0].Subject,
			req.TLS.PeerCertificates[0].SerialNumber,
		)
	})

	httpServer := &http.Server{
		Addr:      ":8080",
		TLSConfig: tlsConfig,
	}

	httpServer.ListenAndServeTLS(
		"/app/certs/server/crt.pem",
		"/app/certs/server/key.pem",
	)
}
```

and the actual testing:

request with the server certificate (for reference) directly towards the dummy endpoint:
```bash
$ curl https://localhost:5447 \
       --cacert certs/ca/crt.pem \
       --cert certs/server/crt.pem --key certs/server/key.pem
cert/subject: CN=server,OU=my-ou,O=my-o,L=my-l,ST=my-st,C=my-c (26161435)
```

using the client certificate which is what we really wanted ..
```bash
$ curl https://localhost:5447 \
       --cacert certs/ca/crt.pem \
       --cert certs/client1/crt.pem --key certs/client1/key.pem
cert/subject: CN=client1,OU=my-ou,O=my-o,L=my-l,ST=my-st,C=my-c (26161436)
```

trying the same client certificate via traefik shows the configured server certificate on the backend:
```bash
$ curl https://localhost:5443 -H 'Host: server' \
       --cacert certs/ca/crt.pem \
       --cert certs/client1/crt.pem --key certs/client1/key.pem
cert/subject: CN=server,OU=my-ou,O=my-o,L=my-l,ST=my-st,C=my-c (26161435)
``` 
i've checked the PR #2324 where this feature originated from, walked through the relevant code path and i wasn't able to see the section where we're actually accessing the clients certificate to pipe it forward. additionally i didn't find a test case which verifies this feature (all test-backends i've found are http based only), so i'm not even sure that ever worked?

or am i simple misreading the docs and therefore expecting something completely different?

**Comment by steffkes** [comment]
Thinking further on this: given the way that ssl works, i can't see how this should work - at all. 

the typical way it works is as proxy in front of the actual application server. using nginx as an example: they allow to pass certificate information as header to the application server - as almost all other tools do (one way or another) which allows the application to decide how to handle the request based on its headers.

Because, once you terminate the ssl connection .. you're basically done. you can't rebuild the certificate information and attach it on the outgoing request (proxy -> backend) .. that's simply not possible because you don't have access to the certificates private key. _you could generate a temporary certificate (if you have access to the CA) and pass that along .. but that would circumvent the idea of client certificates in my opinion - so we're leaving that aside._

[caddy is trying another approach](https://github.com/mholt/caddy/issues/1375), allowing the proxy more control about how to handle the current request:
> Use `Config.VerifyPeerCertificate()` to give site owner more control over authentication with client certificates.
which is kinda interesting but not really feasible for my use case .. i'd like to have the application more/full control (=access to the peer certificate) about how to handle the request and not offloading that decision to some tool earlier in the chain.

**Comment by steffkes** [comment]
ah .. actually, reading [another caddy-ticket ](https://github.com/mholt/caddy/issues/1904#issuecomment-334035348) it seems that the outstanding item on the first ticket is about implementing CRL (Certificate revocation list) checks .. so not really what i was getting from it.

**Comment by kuramal** [comment]
@ldez  when set ingress.kubernetes.io/pass-tls-cert=true, is that mean traefik will forward the end-user client cert to the backend ?

**Comment by jbdoumenjou** [comment]
Hi @steffkes @kuramal,

Thanks for your interest in Træfik.

Unfortunately, this option does not work as expected.
In fact, it activates the TLS between Træfik and the backend with the first certificate defined in the Træfik configuration but it does not pass the client certificate to the backend.
We already analyzed this issue and we'll propose a fix ASAP.

**Comment by steffkes** [comment]
Thanks for confirming @jbdoumenjou - if you need any help further debugging this and/or testing any kind of change/patch, just drop me a note .. i'm happy to help

**Comment by emilevauge** [comment]
To clarify things, as Traefik is not a TCP RP, we cannot provide transparent tls passthrough.

Instead, we plan to implement something similar to what is done [on Nginx](http://nginx.org/en/docs/http/ngx_http_ssl_module.html#var_ssl_client_escaped_cert).
We would like to be able to set the client TLS cert into a specific header forwarded to the backend server. For example, `X-Forwarded-SSL-Client-Cert` would contain the escaped client certificate in PEM.

Of course, this solution need your backend servers to be aware of that.

**Comment by mrdrup** [comment]
Thanks for the update @emilevauge
I think suggested solution is something that everyone expects unless you want to convert Træfik into L3 LB :)

Btw, do you have estimates on the implementation? 
Please contact me if you need any help with testing/debugging. Really happy to help.

**Comment by GKTheOne** [comment]
@emilevauge Would the specific header be configurable? ie, "traefik.frontend.passTLSCertHeader=X-NonStandard-Choice_BAD_DEV" (inherited code)

**Comment by jnovack** [comment]
> Instead, we plan to implement something similar to what can be done with Nginx...  For example, X-Forwarded-SSL-Client-Cert would contain the escaped client certificate in PEM.

Will you be providing DN and expiration as well?  No need to parse the certificate twice (both in Traefik AND on the backend) when Traefik can supply the simple details (like CN and expiry) which most backends are really checking against.

**Comment by kalise** [comment]
@jbdoumenjou 

> In fact, it activates the TLS between Træfik and the backend with the first certificate defined in the Træfik configuration but it does not pass the client certificate to the backend.

What does exactly means? I'm trying to use  Traefik as API gateway in front of multiple remote API servers but, in my case, I need Traefik to establish a mutual TLS authentication between itself and the backend using a different client certificate used for authenticate with the frontend. I was able to setup this with basic auth `traefik.ingress.kubernetes.io/auth-secret` and `traefik.ingress.kubernetes.io/auth-secret` but not able to do the same with the tls auth. It seems `traefik.ingress.kubernetes.io/auth-tls-secret` is not the case or it is not working as espected.

Any idea? 

Thanks

**Comment by jbdoumenjou** [comment]
@kalise, this is another topic. I think that you should be interested in the following proposal: #3038.

**Comment by ldez** [comment]
Closed by #3826
